superset-frontend/src/filters/components/Select/SelectFilterPlugin.tsx

```diff
@@ -53,33 +109,68 @@ export default function PluginFilterSelect(props: PluginFilterSelectProps) {
     inverseSelection,
     inputRef,
     defaultToFirstItem,
+    searchAllOptions,
   } = formData;
 
-  const forceFirstValue =
+  const isDisabled =
     appSection === AppSection.FILTER_CONFIG_MODAL && defaultToFirstItem;
 
   const groupby = ensureIsArray<string>(formData.groupby);
   // Correct initial value for Ant Select
-  const initSelectValue: SelectValue =
-    // `defaultValue` can be `FIRST_VALUE` if `defaultToFirstItem` is checked, so need convert it to correct value for Select
-    defaultValue === FIRST_VALUE ? [] : defaultValue ?? [];
-
-  const firstItem: SelectValue = data[0]
-    ? (groupby.map(col => data[0][col]) as string[]) ?? initSelectValue
-    : initSelectValue;
 
   // If we are in config modal we always need show empty select for `defaultToFirstItem`
-  const [values, setValues] = useState<SelectValue>(
-    defaultToFirstItem && appSection !== AppSection.FILTER_CONFIG_MODAL
-      ? firstItem
-      : initSelectValue,
-  );
+  const [values, setValues] = useState<SelectValue>([]);
   const [currentSuggestionSearch, setCurrentSuggestionSearch] = useState('');
+  const [dataMask, dispatchDataMask] = useReducer<DataMaskReducer>(reducer, {});
+
+  const searchWrapper = (val: string) => {
+    if (searchAllOptions) {
+      debouncedOwnStateFunc(dispatchDataMask, val);
+    }
+    setCurrentSuggestionSearch(val);
+  };
 
   const clearSuggestionSearch = () => {
     setCurrentSuggestionSearch('');
+    if (searchAllOptions) {
+      dispatchDataMask({
+        type: 'ownState',
+        ownState: {
+          search: null,
+        },
+      });
+    }
   };
 
+  useEffect(() => {
+    if (!isDisabled && defaultValue?.length) {
+      // initialize to default value
+      setValues(defaultValue);
+    }
+    // initialize column types (these should only be set once)
+    if (searchAllOptions) {
+      dispatchDataMask({
+        type: 'ownState',
+        ownState: {
+          coltypeMap,
+        },
+      });
+    }
+  }, []);
```

Comment:
I think this logic should be on line 124, when we create reducer in order to avoid unneccessary rerendering